connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -431,4 +436,55 @@ private String trace(Throwable t) {
             return null;
         }
     }
+
+    /*
+     * Performs a reverse transformation on a set of task configs, by replacing values with variable references.
+     * If the raw task config is not available, it will use the raw connector config to obtain variable references.
+     */
+    public static List<Map<String, String>> reverseTransform(String connName,
+                                                             ClusterConfigState configState,
+                                                             List<Map<String, String>> configs) {
+
+        // Find the config keys in the raw connector config that have variable references
+        Map<String, String> rawConnConfig = configState.rawConnectorConfig(connName);
+        Set<String> connKeysWithVariableValues = keysWithVariableValues(rawConnConfig, ConfigTransformer.DEFAULT_PATTERN);
+
+        List<Map<String, String>> result = new ArrayList<>();
+        int index = 0;
+        for (Map<String, String> config : configs) {
+            Map<String, String> newConfig = new HashMap<>(config);
+            Map<String, String> rawConfig = configState.rawTaskConfig(new ConnectorTaskId(connName, index));
+            Set<String> keysWithVariableValues;
+            if (rawConfig != null) {
```

Comment:
I'm a bit confused by the logic here wrt the "raw" task config. That's effectively something we never get from the connector. If we have a config in the ConfigState, we would have generated that from this `reverseTransform` at some point, right?

More specifically, I think there are a couple of cases we would be trying to cover:

1. Most common case where a variable replacement config exists in both the connector config and the task config.
2. Task config contains additional config fields that use variable replacement and don't exist in the connector config
3. A connector is upgraded to a new version and has a new config field that didn't exist in previous versions. It won't exist in the previous versions of the task config either.

1 works if you use either config. 2 seems like it shouldn't need to go through this process because the connector should fill in the config with the variable, not the value. It only doesn't work if you take something that was externalized in the connector config and move it into a different config for the task. (I'll come back to this below.) 3 only works if you use the raw connector config since on upgrade the cluster state would have an entry for the task config already but it wouldn't include that new, additional config.

More generally, we should probably discuss if this is the solution we want. Coming back to the issue with 2, this whole solution is based on the assumption that the connector and task configs match. This is definitely a new assumption and not enforced in any way by the framework. I think this is probably a reasonable assumption to make since all the connectors I've seen work like this since it is easier, but we've basically never enforced anything about the task configs. 2 is also kind of interesting case anyway because to do this successfully you'd really need the untransformed config as well.

Reply:
@ewencp , I really appreciate your thoughtful feedback.  

Before I try to take in your analysis above, I just wanted to point out (unless I'm missing something), that I think `ConnectorsResource.putTaskConfigs` allows one to go through the REST API to modify the task configs in the `ConfigBackingStore`, which then gets passed to the `ClusterConfigState` constructor.  This is an instance where the task configs would not go through `reverseTransform`.  So that is one reason I thought I would have to use the "raw" task config to perform the proper reverse transformation, as the user may have inserted a variable directly using the REST API.  Please correct me if I am misunderstanding something.

Reply:
Ok, after thinking more, I see your point.  It seems that `reverseTransform` is only called on the result of `worker.connectTaskConfigs`, which is supposed to represent the newest task configs.  So in that case the old task configs can be ignored, even if they contain variable references.  So it seems that perhaps we should only check the "raw" connector config, as you suggested.  Of course, as you pointed out, this is based on the assumption that any keys that would be reverse transformed match between the connector and the task.  Without this assumption, we would need some other information to indicate config lineage between task and connector, which we currently don't have.